libraries/chain/combined_database.cpp

```diff
@@ -0,0 +1,470 @@
+         if (std::is_same<value_t, database_header_object>::value) {
+            return;
+         }
+
+         // skip the kv_db_config as it only determines where the kv-database is stored
+         if (std::is_same_v<value_t, kv_db_config_object>) {
+            return;
+         }
+
+#warning TODO: Revisit this and adpat to new design
+         if constexpr (std::is_same_v<value_t, kv_object>) {
+            snapshot->write_section<value_t>([this](auto& section) {
+               // This ordering depends on the fact the eosio.kvdisk is before eosio.kvram and only eosio.kvdisk can be
+               // stored in rocksdb.
+               if (db.get<kv_db_config_object>().backing_store == backing_store_type::ROCKSDB) {
+                  std::unique_ptr<rocksdb::Iterator> it{ kv_database.rdb->NewIterator(rocksdb::ReadOptions()) };
+                  std::vector<char>                  prefix = rocksdb_contract_kv_prefix;
+                  b1::chain_kv::append_key(prefix, kvdisk_id.to_uint64_t());
+                  it->Seek(b1::chain_kv::to_slice(rocksdb_contract_kv_prefix));
+                  while (it->Valid()) {
+                     auto key = it->key();
+                     if (key.size() < rocksdb_contract_kv_prefix.size() ||
+                         !std::equal(prefix.begin(), prefix.end(), key.data()))
+                        break;
+                     uint64_t    contract;
+                     std::size_t key_prefix_size = prefix.size() + sizeof(contract);
+                     EOS_ASSERT(key.size() >= key_prefix_size, database_exception, "Unexpected key in rocksdb");
+                     std::reverse_copy(key.data() + prefix.size(), key.data() + key_prefix_size,
+                                       reinterpret_cast<char*>(&contract));
+                     auto           value = it->value();
+                     kv_object_view row{ kvdisk_id,
+                                         name(contract),
+                                         { { key.data() + key_prefix_size, key.data() + key.size() } },
+                                         { { value.data(), value.data() + value.size() } } };
+                     section.add_row(row, db);
+                     it->Next();
+                  }
+               }
+               decltype(utils)::template walk<by_kv_key>(
+                     db, [this, &section](const auto& row) { section.add_row(row, db); });
+            });
+            return;
+         }
+
+         snapshot->write_section<value_t>([this](auto& section) {
+            decltype(utils)::walk(db, [this, &section](const auto& row) { section.add_row(row, db); });
+         });
+      });
+
+      add_contract_tables_to_snapshot(snapshot);
+
+      authorization.add_to_snapshot(snapshot);
+      resource_limits.add_to_snapshot(snapshot);
+   }
+
+   void combined_database::read_from_snapshot(const snapshot_reader_ptr& snapshot, uint32_t blog_start,
```

Comment:
Same readability concerns as `add_to_snapshot`.

Reply:
`clang-format` hasn't been configured properly. I don't know what to do about it.

Reply:
Manually edit it. I don't think many people on our team are relying on clang-format.

Reply:
I changed the formatting manually.